src/ipvs/ip_vs_synproxy.c

```diff
@@ -947,6 +947,181 @@ static int syn_proxy_send_rs_syn(int af, const struct tcphdr *th,
     return EDPVS_OK;
 }
 
+/* Reuse mbuf and construct TCP RST packet */
+static int syn_proxy_build_tcp_rst(int af, struct rte_mbuf *mbuf,
+                                   void *iph, struct tcphdr *th,
+                                   uint32_t l3_len, uint32_t l4_len,
+                                   uint32_t seq) {
+    struct netif_port *dev;
+    uint16_t tmpport;
+    uint16_t tcph_len, payload_len;
+    struct iphdr *ip4h;
+    struct ip6_hdr *ip6h;
+
+    if (unlikely(l4_len < sizeof(struct tcphdr)))
+        return EDPVS_INVPKT;
+
+    tcph_len = th->doff * 4;
+
+    if (unlikely(l4_len < tcph_len))
+        return EDPVS_INVPKT;
+
+    payload_len = l4_len - tcph_len;
+
+    /* set tx offload flags */
+    dev = netif_port_get(mbuf->port);
+    if (unlikely(!dev)) {
+        RTE_LOG(ERR, IPVS, "%s: device eth%d not found\n",
+                __func__, mbuf->port);
+        return EDPVS_NOPROT;
+    }
+    if (likely(dev && (dev->flag & NETIF_PORT_FLAG_TX_TCP_CSUM_OFFLOAD))) {
+        if (af == AF_INET6)
+            mbuf->ol_flags |= (PKT_TX_TCP_CKSUM | PKT_TX_IPV6);
+        else
+            mbuf->ol_flags |= (PKT_TX_TCP_CKSUM | PKT_TX_IP_CKSUM | PKT_TX_IPV4);
+    }
+
+    /* exchange ports */
+    tmpport = th->dest;
+    th->dest = th->source;
+    th->source = tmpport;
+    /* set window size to zero */
+    th->window = 0;
+    /* set seq and ack_seq */
+    if (th->syn)
+        th->ack_seq = htonl(ntohl(th->seq) + 1);
+    else
+        th->ack_seq = htonl(ntohl(th->seq) + payload_len);
+    th->seq = seq;
+    /* set TCP flags */
+    th->th_flags = 0;
+    th->rst = 1;
+    th->ack = 1;
+
+    /* truncate packet if TCP payload presents */
+    if (payload_len > 0) {
+        if (rte_pktmbuf_trim(mbuf, payload_len) != 0) {
+            return EDPVS_INVPKT;
+        }
+    }
+
+    if (AF_INET6 == af) {
+        struct in6_addr tmpaddr;
+        ip6h = iph;
+
+        tmpaddr = ip6h->ip6_src;
+        ip6h->ip6_src = ip6h->ip6_dst;
+        ip6h->ip6_dst = tmpaddr;
+        ip6h->ip6_hlim = 63;
+        ip6h->ip6_plen = htons(ntohs(ip6h->ip6_plen) - payload_len);
+
+        /* compute checksum */
+        if (likely(mbuf->ol_flags & PKT_TX_TCP_CKSUM)) {
+            mbuf->l3_len = l3_len;
+            mbuf->l4_len = l4_len;
```

Comment:
`mbuf->l4_len` is L4 header length.

Reply:
TCP payload has been truncated and `l4_len` is equal to TCP header length. Please see line 1009.